substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageBFDNameProvider.java

```diff
@@ -383,14 +420,166 @@ public void setNativeLibs(NativeLibraries nativeLibs) {
+         * the mangler can embed the short symbol instead of generating the previously mangled full
+         * text. Short symbol names are bound in a well-defined sequence at well-defined points
+         * during mangling. This allows the demangler to identify exactly which text in the previous
+         * input must be used to replace a short symbol name.
+         *
+         * A binding occurs whenever a simple name is mangled outside of a namespace. For example, a
+         * top level class name mangled as 3Hello will bind the next available short name with the
+         * result that it will demangle to Hello and can be used to replace later occurrences of the
+         * string Hello.
+         *
+         * When a sequence of simple names is mangled inside a namespace substitution bindings are
+         * recorded for each successive composite namespace prefix but not for the final symbol
+         * itself. For example, when method name Hello::main is mangled to N5Hello4main4 a single
+         * binding is recorded which demangles to Hello. If a class has been loaded by an
+         * application loader and, hence, has a name which includes a loader namespace prefix (to
+         * avoid the possibility of the same named class being loaded by two different loaders) then
+         * the method name, say AppCL504::Hello::main, would mangle to a namespace encoding with 3
+         * elements N8AppCL5045Hello4mainE which would introduce two bindings, the first demangling
+         * to AppCL504 and the second to AppCL504::Hello.
+         *
+         * A binding is also recorded whenever a pointer type is mangled. The bound symbol demangles
+         * to whatever the text decoded from the scope of the P followed by a '*' to translate it to
+         * a pointer type. For example, when the type in parameter list (Foo*) is encoded as P3Foo a
+         * binding is recorded for the pointer as well as for the type Foo. The first binding
+         * demangles to Foo. The second binding demangles to Foo*.
+         *
+         * n.b. repeated use of the pointer operator results in repeated bindings. So, if a
+         * parameter with type Foo** were to be encoded as PP3Foo three bindings would recorded
+         * which demangle to Foo, Foo* and Foo**. However, multiple indirections do not occur in
+         * Java signatures.
+         *
+         * n.b.b. repeated mentions of the same symbol would redundantly record a binding. For
+         * example if int Hello::compareTo(Hello*) were mangled to _ZN5Hello9compareToEJiP5Hello the
+         * resulting bindings would be S_ ==> Hello, S0_ ==> Hello::compareTo, S1_ ==> Hello and S2_
+         * ==> Hello* i.e. both S_ and S1_ would demangle to Hello. This situation should never
+         * arise. A name can always be correctly encoded without repeats. In the above example that
+         * woudl be _ZN5Hello9compareToEJiPS_.
+         */
```

Comment:
Perhaps it's worth mentioning that the above are based on https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling-compression and make clear that GraalVM doesn't really create any bindings/lookup-tables in the debug info sections.

Reply:
Unfortunately, it is not based on that document. That document merely reflects some aspects of the implementation (and falls a little short as a specification). The correct reference would be to a GNU g++ spec (which I have not yet found) or, in its absence, to the BFD source code (which can be found in both the GNU binutils and gdb trees).

I don't think there is a need for any specific reference. The gdb source tree is definitive for this and many other aspects of the way the Linux debug info generator works and really needs to be read to understand and verify many details of the implementation. I summarized the relevant points here rather than referenced the gdb code because we only need to consider a subset of the encoding scheme when it comes to Java i.e. this is just a TL;DR for the code.

Reply:
It seems the only GNU's extension is the `J` for return type, other part follows the itanium's spec (not sure, at least respected by llvm and gcc)

https://github.com/gcc-mirror/gcc/blob/12039c9f40a81605e59951f80007827bc0413573/libiberty/cp-demangle.c#L3059

I also didn't find llvm's c++filt supporting mangling return type, does that mean debug information on MacOS is infeasible?

Reply:
The BFD name mangling support is implemented for Linux in order support debugging using gdb and use of Linux analysis tool such as perf and valgrind. These programs expect to see return types in mangled names. I don't know what LLVM expects on Linux. If it cannot recognize return types then it is goign to have a problem debugging librraies generated by gcc/g++.

Debug info generation is not currently available for MacOS. So, whether the current code works correctly on MacOS is not an issue. If we want to reuse some of the current debug generator code on MacOS and that requires name mangling, for llvm or for other tools, then it may be necessary to implement a new mangler. If MacOS tools will accept GNUv3 mangled names without return types then it might be possible to customize the current mangler to cater for that. However, I believe there are cases (synthetic bridge methods) where two distinct Java methods are only differentiated by return type. In that case omitting the return type would map both methods to the same mangled name. So, that problme amy need ot be finessed.